Implemented Moving Histogram algorithm

spran180 opened the pull request:
Fixes #752

Summary  
This PR implemented Moving Histogram Algorithm as stated in #752. This change includes the addition of a new class 'MovHistogram' and its associated unit tests.

gavv commented:
Hi, I see it's draft and haven't look at the code closely, but here a couple of preliminary comments:

* PRs should be targeted to develop branch, not master (see [here](https://roc-streaming.org/toolkit/docs/development/contribution_guidelines.html))
* Please follow code style (see [here](https://roc-streaming.org/toolkit/docs/development/coding_guidelines.html#coding-style))
* Template classes should be defined in header, not in .cpp

spran180 commented:
Hi @gavv,

Thanks for reviewing my code. I will correct it shortly as per your suggestions.

spran180 commented:
Hello @gavv,

I have addressed the issues you highlighted. The branch has been rebased to develop, and the code has been revised to align with the coding style. Everything should be in order now. Kindly inform me if any further changes or checks are needed.

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,100 @@
+        : value_range_(value_range)
+        , num_bins_(num_bins)
+        , window_length_(window_length)
+        , bin_width(static_cast<T>(value_range) / static_cast<T>(num_bins))
```
I suppose you have  wrong understanding of what `value_range` is: it is supposed to be a pair of values -- $`(min, max)`$, not just a single value. So each $`bin_i`$ $`\left(i \in \left[0,N\right)\right)`$ will be used for counting values in range
$$\left[ min + \dfrac{i(max-min)}{N}, min + \dfrac{(i+1)(max-min)}{N} \right).$$

In your case a range is always a $`\left[0, max\right)`$ which means a loss of generality.

@gavv will correct me if I am wrong.

gavv replied:
I agree, I also had this in mind.

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,100 @@
+        , ring_buffer_(arena, window_length)
+        , bins_(arena)
+        , valid_(false) {
+        if (num_bins == 0 || window_length_ == 0) {
```
it would be great not to mix parameters and member values in a single expression

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    //! returns the number of values in the bin
+    int getBinCounter(size_t bin_index) const {
```
style: we do not use CamelCase for function naming for C language anywhere in this project; please use underscores instead

gavv replied:
And also please add trailing underscore to private methods and fields.

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,100 @@
+    size_t window_length_;
+    T bin_width;
+    RingQueue<T> ring_buffer_;
+    Array<int> bins_;
```
IIUC, here we should use `T` instead of `int`

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
I didn't look into the `add_value()` thoroughly because it is going to be fixed according to other comments. But it has some problems too. Please, read the task definition, especially these 2 points:

> - increment counter of the bin corresponding to the added value
> - decrement counter of the bin corresponding to the removed value

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,100 @@
+    //! returns the bin index for a given value
+    int getBinIndex(const T& value) const {
+        if (value > value_range_) {
+            return -1;
```
Consider validating `value` in `add_value()` instead of using a placeholder here. Therefore the returning type will be size_t.

I asked Victor, he told me you can use saturation there:
- if value > max => value = max
- if value < min => value = min

gavv commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,100 @@
+        , bins_(arena)
+        , valid_(false) {
+        if (num_bins == 0 || window_length_ == 0) {
+            roc_panic(
```
We should do this check before dividing by num_bins.

gavv commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,100 @@
+    //! Add a value to the histogram.
+    void add_value(const T& value) {
+        if (ring_buffer_.size() == window_length_) {
+            T oldest_bin_ = ring_buffer_.front();
```
I think oldest_value would be a better name? Also the var can be const.

gavv reviewed (commented):
Thanks for PR! A couple of comments from me as well.

spran180 commented:
Thank you for the review, @adrianopol and @gavv. I will look into it and update the code accordingly.

spran180 commented:
Hey @adrianopol, I'm using `pair<T, T>` to represent the value range `{min_val, max_val}` instead of creating separate variables. Is this approach correct?

gavv commented:
> Hey @adrianopol, I'm using `pair<T, T>` to represent the `value_range` `{min_val, max_val}` instead of creating separate variables. Is this approach correct?

We don't use STL containers, so please use separate arguments. See here: https://roc-streaming.org/toolkit/docs/development/coding_guidelines.html

spran180 commented:
Hey @gavv / @adrianopol, 

I've addressed the changes that you've requested pls go through it once and let me know if any further changes are required.

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,111 @@
+            T oldest_value = ring_buffer_.front();
+            ring_buffer_.pop_front();
+            size_t oldest_bin_index = get_bin_index_(oldest_value);
+            if (oldest_bin_index < num_bins_) {
```
Is there any case when this `if` returns false? If not, you may drop it.

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,111 @@
+private:
+    //! returns the bin index for a given value
+    size_t get_bin_index_(const T& value) const {
+        return static_cast<size_t>((value - value_range_min_) / bin_width_);
```
There is a corner case (see the right parenthesis my previous comment which mean a non-inclusive range: https://github.com/roc-streaming/roc-toolkit/pull/760#discussion_r1684971776): if `value == value_range_max`, bin index will be `num_bins_` (which means N+1), which is incorrect.

There are multiple solutions how to deal with it. One is to add a branch:
```
if (value == value_range_max_) {
    return num_bins_ - 1;
}
return ...
```

This should fix the problem in line 75. Also it would be nice to add a test for this case.

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    //! returns the number of values in the bin
```
```suggestion
    //! Get the number of values in the given bin.
```

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,111 @@
+    }
+
+private:
+    //! returns the bin index for a given value
```
```suggestion
    //! Get the bin index for the given value.
```

spran180 commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,123 @@
+                      "than 0 and value_range_min must be less than value_range_max");
+        }
+
+        bin_width_ = (value_range_max - value_range_min) / static_cast<T>(num_bins);
```
Hey @adrianopol,

Since `num_bins` is initialized as `size_t`, it needs to be cast to type `T`. Please let me know if this is incorrect.

adrianopol replied:
This is correct, I just mean that C/C++ already do this implicit promotion (see https://en.cppreference.com/w/cpp/language/implicit_conversion): if at least on of operands in an arithmetic expression have type float (or double) then the whole expression will have type float (or double).

If T = float, then without casting we have `(float - float) / size_t` == `float / size_t` == `float`.

If T = `short`, we have `(short - short) / size_t` == `short / size_t` == `size_t`. Value (max - min) is always positive, so we have no problem with loosing sign here.

You may leave your variant as is, it is correct too.

gavv replied:
If T is `short`, then `(value_range_max - value_range_min) / num_bins` is size_t, and assigning size_t to bin_width_ would trigger a warning.

If T is `ssize_t` (signed version of size_t), then `(value_range_max - value_range_min) / num_bins` would trigger a warning.

So I think we need to keep the cast here.

BTW we usually use c-style casts or constructor-style casts instead of static_cast, but that's not a big deal.

adrianopol commented on src/tests/roc_core/test_mov_histogram.cpp:
```diff
@@ -0,0 +1,109 @@
+                              win_length);
+    CHECK(hist.is_valid());
+
+    size_t values[win_length];
```
What is the purpose of this value? I think, you can eliminate it here and in the second test.

spran180 replied:
The `values` array used to store the added values that we add to the Histogram, but I think it's better to remove it.

adrianopol commented on src/internal_modules/roc_core/mov_histogram.h:
```diff
@@ -0,0 +1,123 @@
+
+    //! Get the number of values in the given bin.
+    size_t get_bin_counter(size_t bin_index) const {
+        return bins_[bin_index];
```
if `bin_index` > `num_bins_` we should panic here

gavv replied:
Actually `Array` already has bound checking.

adrianopol replied:
ah, OK

adrianopol reviewed (changes requested):
@spran180 , thank you for implementation.

We also need more test cases here, for example:
- T = float
- min and max < 0
- win_length == 1
They may be added in a different PR if you wish.

gavv commented:
> @spran180 , thank you for implementation.
> 
> We also need more test cases here, for example:
> 
>     * T = float
> 
>     * min and max < 0
> 
>     * win_length == 1
>       They may be added in a different PR if you wish.

I agree.

Also, current tests don't cover a few important aspects:

* when there are multiple values in every bin (currently every bin is either 0 or 1)
* how bin counters increment and decrement when rolling window is moved (e.g. you call add_value, some bin changes from N to N-1, and another bin changes from K to K+1) 
* out of bounds values (clamping)

@spran180 Let me know if you would like to add these tests or it's better to create a new issue for that.

spran180 commented:
> > @spran180 , thank you for implementation.
> > We also need more test cases here, for example:
> > ```
> > * T = float
> > 
> > * min and max < 0
> > 
> > * win_length == 1
> >   They may be added in a different PR if you wish.
> > ```
> 
> I agree.
> 
> Also, current tests don't cover a few important aspects:
> 
> * when there are multiple values in every bin (currently every bin is either 0 or 1)
> * how bin counters increment and decrement when rolling window is moved (e.g. you call add_value, some bin changes from N to N-1, and another bin changes from K to K+1)
> * out of bounds values (clamping)
> 
> @spran180 Let me know if you would like to add these tests or it's better to create a new issue for that.

Yes, I believe I should create a separate PR to add tests.

spran180 commented:
Hey @gavv / @adrianopol,

Could you please guide me on how to pass the `build / linux-x86_64/ubuntu-24.04 (pull_request)` test that is failing after 6 seconds?

gavv commented:
The build failure was docker hub timeout, not related to the PR. I've restarted the build and it's gone.

> Yes, I believe I should create a separate PR to add tests.

@spran180 Great, do you plan to work on it, or do I create an issue for the future?

spran180 commented:
> The build failure was docker hub timeout, not related to the PR. I've restarted the build and it's gone.
> 
> > Yes, I believe I should create a separate PR to add tests.
> 
> @spran180 Great, do you plan to work on it, or do I create an issue for the future?

Indeed, I am interested in working on it.

gavv commented:
Thanks for PR!

> Indeed, I am interested in working on it.

Perfect. I've copied Andrew's and mine comments here: https://github.com/roc-streaming/roc-toolkit/issues/752#issuecomment-2254591934
